### PR TITLE
Fix Pro generator multiline and template-literal rewrites

### PR DESCRIPTION
**Pull request by justin808**
### Summary

Improve the Pro install generator rewrite logic so multiline non-parenthesized `gem "react_on_rails"` declarations keep trailing options while replacing the base gem entry. Harden module-specifier rewriting around template literals by preserving escaped sequences and correctly detecting multiline template-literal starts after a closed inline template.

### Pull Request checklist

- [x] Add/update test to cover these changes
- [x] ~Update documentation~
- [x] ~Update CHANGELOG file~

### Other Information

Validated with `bundle exec rspec react_on_rails/spec/react_on_rails/generators/pro_generator_spec.rb:219 react_on_rails/spec/react_on_rails/generators/pro_generator_spec.rb:1007 react_on_rails/spec/react_on_rails/generators/pro_generator_spec.rb:1019`.

---

> [!NOTE]
> **Medium Risk**
> Touches the Pro install generator’s Gemfile and JS source rewriting logic; subtle parsing/state bugs could cause incorrect dependency swaps or missed/over-eager import rewrites in real-world edge cases.
> 
> **Overview**
> Improves the Pro generator’s Gemfile swap to correctly consume *non-parenthesized multiline* `gem "react_on_rails"` declarations and carry forward trailing options/guards while stripping only version constraints, avoiding orphaned lines and stray/trailing commas.
> 
> Hardens JS module-specifier rewriting around template literals by restoring inline template placeholders without altering escaped sequences, and by more accurately detecting multiline template-literal boundaries across inline literals and block-comment transitions (so imports between these regions are rewritten correctly). Adds focused specs for these multiline Gemfile and template-literal edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 5a6e361f2001539a0239a259282b835d5e755636. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Preserve non-version gem options during upgrade without leaving trailing commas, extra whitespace, or blank-line artifacts.
  * More robust template-literal handling across inline and multiline cases, including escaped sequences, correct backtick detection, and block-comment boundary transitions.

* **Tests**
  * Added unit tests for non-parenthesized multiline gem declarations.
  * Added tests for template-literal edge cases: placeholder restoration, imports between literals, and comment-boundary scenarios.

### CODE REVIEW

**Comment by justin808** [comment]
@claude review this PR.

File: react_on_rails/lib/generators/react_on_rails/pro_generator.rb [hunk 1]
```diff
@@ -483,7 +480,8 @@ def rewrite_non_comment_lines(content)
                 rewrite_line_before_template_literal_open(
                   line,
                   pending_multiline_module_call_depth,
-                  pending_multiline_static_import_specifier
+                  pending_multiline_static_import_specifier,
+                  in_block_comment: in_block_comment
                 ) { |line_fragment| yield line_fragment }
               in_multiline_template_literal = updated_template_literal_state
               in_block_comment = true if unclosed_block_comment_starts?(rewritten_line)
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/generators/react_on_rails/pro_generator.rb]: Fixed in 5a6e361. Changed both lines 476 and 487 from conditional (`= true if`) to unconditional (`=`) assignment so `in_block_comment` is properly reset when a block comment closes on the same line a template literal opens. Also added an extended test that verifies imports on lines after the template literal closes are still rewritten correctly.

File: react_on_rails/lib/generators/react_on_rails/pro_generator.rb [hunk 2]
```diff
@@ -483,7 +480,8 @@ def rewrite_non_comment_lines(content)
                 rewrite_line_before_template_literal_open(
                   line,
                   pending_multiline_module_call_depth,
-                  pending_multiline_static_import_specifier
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/generators/react_on_rails/pro_generator.rb:486]: Fixed in 5a6e361 — same root cause as the claude[bot] comment. Both lines 476 and 487 now use unconditional assignment: `in_block_comment = unclosed_block_comment_starts?(rewritten_line)`.

File: react_on_rails/spec/react_on_rails/generators/pro_generator_spec.rb [hunk 3]
```diff
@@ -986,6 +1005,39 @@
...
+
+    it "rewrites imports after a block comment closes before a multiline template literal start" do
+      source = <<~JS
+        /*
+         odd backtick ` still comment */ import("react-on-rails"); `multiline-start
+      JS
+
+      rewritten = generator.send(:rewrite_react_on_rails_module_specifiers, source)
+
+      expect(rewritten).to include('odd backtick ` still comment */ import("react-on-rails-pro"); `multiline-start')
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/spec/react_on_rails/generators/pro_generator_spec.rb:1038]: Added in 5a6e361. The new test "rewrites imports after a block-comment-to-template-literal transition closes" verifies that imports on lines after the template literal closes are still rewritten, which exercises the `in_block_comment` reset fix.

File: react_on_rails/lib/generators/react_on_rails/pro_generator.rb [hunk 4]
```diff
@@ -757,17 +760,54 @@ def rewrite_line_before_template_literal_open(line, pending_depth, pending_multi
...
+            scan_index += 1
+            next
+          end
+
+          break if line[scan_index, 2] == "//"
+
+          if line[scan_index, 2] == "/*"
+            in_block_comment = true
+            scan_index += 2
+            next
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/generators/react_on_rails/pro_generator.rb]: Fixed in 5a6e361. Added a `skip_comments:` parameter to `unescaped_backtick_indexes`. `first_unescaped_backtick_index` (used when scanning for the closing backtick inside a template literal) now passes `skip_comments: true` so `/*` and `//` in template literal content are correctly treated as literal text, not comment tokens. Added a test covering this edge case.

File: react_on_rails/lib/generators/react_on_rails/pro_generator.rb [hunk 5]
```diff
@@ -757,17 +760,54 @@ def rewrite_line_before_template_literal_open(line, pending_depth, pending_multi
       end
 
       def first_unescaped_backtick_index(line)
+        unescaped_backtick_indexes(line).first
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/generators/react_on_rails/pro_generator.rb]: Fixed in 5a6e361 — same issue as the chatgpt-codex-connector comment. `first_unescaped_backtick_index` now passes `skip_comments: true` to `unescaped_backtick_indexes` so comment tokens inside template literal content are ignored.

File: react_on_rails/lib/generators/react_on_rails/pro_generator.rb [hunk 6]
```diff
@@ -757,16 +754,29 @@ def rewrite_line_before_template_literal_open(line, pending_depth, pending_multi
...
+      end
+
+      def opening_backtick_index_for_multiline_start(line)
+        backtick_indexes = unescaped_backtick_indexes(line)
+        return nil if backtick_indexes.empty? || backtick_indexes.length.even?
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/generators/react_on_rails/pro_generator.rb]: The `positive?` guard is intentional — when `opening_index` is 0 the backtick is at position 0, so `line_prefix = line[0, 0]` is `""` and there is nothing before the backtick to rewrite. The tracking calls on an empty prefix are no-ops, so the early return is safe. Adding a comment explaining this would be a style change outside the scope of this PR.

File: react_on_rails/spec/react_on_rails/generators/pro_generator_spec.rb [hunk 7]
```diff
@@ -986,6 +1004,28 @@
...
+      source = <<~'JS'
+        const inlineTemplate = `\\n and \\1`;
+        const ror = require("react-on-rails");
+      JS
+
+      rewritten = generator.send(:rewrite_react_on_rails_module_specifiers, source)
+
+      expect(rewritten).to include('const inlineTemplate = `\\\\n and \\\\1`;')
+      expect(rewritten).to include('const ror = require("react-on-rails-pro");')
+    end
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/spec/react_on_rails/generators/pro_generator_spec.rb:1018]: This analysis is incorrect. Single-quoted heredocs (`<<~'JS'`) perform **no escape processing at all** — unlike single-quoted strings where `\\\\` → `\\`. So `\\\\n` in the heredoc is literally two backslashes + n in memory. The expectation `'\\\\\\\\n'` (four source backslashes in a single-quoted string) also evaluates to two backslashes + n in memory (`\\\\` → `\\` twice). Input and expectation match. The test passes as-is and the assertion is correct.

File: react_on_rails/lib/generators/react_on_rails/pro_generator.rb [hunk 8]
```diff
@@ -830,13 +870,36 @@ def match_multiline_parenthesized_base_gem(lines, start_index)
...
+
       def build_pro_gem_replacement_line(indentation:, quote:, suffix:, parenthesized_gem_call: false)
         normalized_suffix = suffix || "\n"
         normalized_suffix = "#{normalized_suffix}\n" unless normalized_suffix.end_with?("\n")
         version_arg_pattern = /\A(?<prefix>\s*,(?:\s*#.*\n|\s++)*)["'][^"']*["'](?<trailing_comma>\s*,)?/
         loop do
           updated_suffix = normalized_suffix.sub(version_arg_pattern) do
-            Regexp.last_match[:trailing_comma] ? Regexp.last_match[:prefix] : ""
+            if Regexp.last_match[:trailing_comma]
+              Regexp.last_match[:prefix].sub(/\n[ \t]*\z/, "")
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/generators/react_on_rails/pro_generator.rb:901]: Investigated — the claim that the normalization regex at line 908 (`/\A,\s*(?:#[^\n]*)?\n\z/`) would consume options across newlines is incorrect. The `(?:#[^\n]*)?` group stops at the first newline, and `\n\z` requires the string to end immediately after. Tested with `, # note\n  require: false\n` — the regex does not match, so `require: false` is preserved. No code change needed.

File: react_on_rails/lib/generators/react_on_rails/pro_generator.rb [hunk 9]
```diff
@@ -648,7 +646,7 @@ def rewrite_outside_inline_template_literals(line)
 
         rewritten_line = yield line_without_inline_templates
         template_placeholders.each do |placeholder, template_literal|
-          rewritten_line = rewritten_line.sub(placeholder, template_literal)
+          rewritten_line = rewritten_line.sub(placeholder) { template_literal }
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/generators/react_on_rails/pro_generator.rb:649]: Thanks for confirming the block-form `sub` fix.

File: react_on_rails/lib/generators/react_on_rails/pro_generator.rb [hunk 10]
```diff
@@ -757,17 +760,54 @@ def rewrite_line_before_template_literal_open(line, pending_depth, pending_multi
       end
 
       def first_unescaped_backtick_index(line)
+        unescaped_backtick_indexes(line).first
+      end
+
+      def opening_backtick_index_for_multiline_start(line, in_block_comment: false)
+        backtick_indexes = unescaped_backtick_indexes(line, in_block_comment: in_block_comment)
+        return nil if backtick_indexes.empty? || backtick_indexes.length.even?
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/generators/react_on_rails/pro_generator.rb:768]: Good observation about the raw-line dependency. The current call site is correct and this is a narrow internal method, so adding a comment would be scope creep for this PR.

File: react_on_rails/lib/generators/react_on_rails/pro_generator.rb [hunk 11]
```diff
@@ -830,13 +870,36 @@ def match_multiline_parenthesized_base_gem(lines, start_index)
...
+        while line_index < lines.length &&
+              line_continues_with_comma?(current_line) &&
+              gem_declaration_continues_on_next_line?(lines[line_index])
+          next_line = lines[line_index]
+          declaration_lines << next_line
+          current_line = next_line unless comment_or_blank_line?(next_line)
+          line_index += 1
+        end
+
+        trailing_suffix = lines[start_index][match_end..].to_s + declaration_lines.drop(1).join
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/generators/react_on_rails/pro_generator.rb:890]: Acknowledged — the dual-purpose usage (build suffix vs. advance index) is intentional. Adding a clarifying comment would be a style change outside this PR's scope.

**Comment by justin808** [comment]
## Review comment triage & resolution (5a6e361)

Triaged 11 review comments across 5 bots. Two real bugs were fixed, two claims were investigated and found incorrect, and documentation nits were declined as out of scope.

### Fixed (2 bugs)

1. **Stale `in_block_comment` flag** (claude[bot], cursor[bot]) — When a block comment closed on the same line a template literal opened, `in_block_comment` was never reset to `false`. After the template literal closed, all subsequent lines were silently skipped. Fix: unconditional assignment on lines 476/487.

2. **`/*` inside template literals treated as block comment** (chatgpt-codex-connector[bot], cursor[bot]) — `first_unescaped_backtick_index` (used to find the closing backtick inside a template literal) incorrectly treated `/*` in template content as a block comment start. Fix: added `skip_comments: true` parameter.

Both fixes include regression tests. All 79 specs pass.

### Investigated & declined (2 false positives)

- **Backslash count off-by-one** (greptile-apps[bot] P1) — Incorrect. Single-quoted heredocs (`<<~'JS'`) perform no escape processing, so `\\n` is two backslashes + n in memory, matching the expectation.
- **Version stripping drops options** (chatgpt-codex-connector[bot] P1) — Incorrect. The normalization regex `\A,\s*(?:#[^\n]*)?\n\z` cannot match multi-line suffixes — `(?:#[^\n]*)?\n\z` limits it to single-line trailing commas. Verified with a Ruby test.

### Declined as out of scope (4 style/documentation nits)

- `positive?` guard comment, parity logic documentation, dual-purpose method comment, and the duplicate threads — all acknowledged with rationale replies.

All 11 review threads are now resolved.
